[qt5-base] Move 'gui' and 'icu' out of 'core' feature

dg0yt opened the pull request:
- #### What does your PR fix?  
  This PR creates separate `gui` and `icu` features, reducing the number of dependencies and the build time for the `core` feature.
  - Adds an `icu` opt-in feature, reducing build time and deployment size of the default configuration.
    (It might make sense to leave it as a default feature to avoid a change for currents users.)
  - Adds a `gui` opt-out feature to `qt5-base` and `qt5-tools`, meant to benefit cross builds which have a host dependency on `qt5-base[core]` or `qt5-tools[core]`.
    (This PR is separated out of work to fix building qt5 for mingw, but android and other triplets might benefit as well.)
  - Adds a `quick` opt-out feature to `qt5-declarative`, meant to benefit users which only need Qt Widgets, not Qt Quick.
 
  In addition, there is some minor cleanup.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

dg0yt committed:
Move gui out of qt5-base[core]

dg0yt commented:
> * This should be beneficial for cross builds which have a host dependency on `qt5-base[core]`.

... if vcpkg wouldn't install default features for host dependencies: https://github.com/microsoft/vcpkg-tool/pull/177.

Osyotr commented on ports/qt5-base/vcpkg.json:
```diff
@@ -1,29 +1,16 @@
-    "freetype",
-    "harfbuzz",
     {
       "name": "icu",
```
ICU can also be moved into a separate/`gui` feature.

dg0yt replied:
Not sure how much work we want to put in Qt5. But I admit that I still use Qt5 without icu (and openssl) because I don't want to deploy it.

Osyotr replied:
`Elapsed time for package icu:x64-windows: 5.296 min`
This would be a significant improvement IMHO

dg0yt committed:
Move icu out of qt5-base[core]

dg0yt committed:
Parse keywords in qt_submodule_installation

The unparsed arguments must not be blindly passed to
qt_download_submodule and qt_build_submodule.

dg0yt committed:
qt5-doc doesn't install anything

dg0yt committed:
Reduce dependencies on qt5-base[gui]

dg0yt committed:
Fix formatting

dg0yt committed:
Drop unused variables

JackBoosY commented:
cc @Neumann-A

dg0yt commented:
It is basically good enough to review, although downstream ports will need changes to fully leverage qt5-base feature control.
In any case, this will be the base for the mingw (or more general: cross-build) changes which are another PR.

Neumann-A commented on ports/qt5-tools/portfile.cmake:
```diff
@@ -1,9 +1,33 @@
+vcpkg_list(SET libs quick widgets)
```
used for ?

Neumann-A commented on ports/qt5-tools/vcpkg.json:
```diff
@@ -11,7 +12,24 @@
+  "default-features": [
+    "gui"
+  ],
+  "features": {
+    "gui": {
+      "description": "Build GUI features, including QWidgets",
+      "dependencies": [
+        {
+          "name": "qt5-base",
+          "default-features": false,
+          "features": [
+            "gui"
+          ]
+        },
+        "qt5-declarative"
+      ]
+    }
+  }
```
need to be finer grained/renamed. Doesn't make sense for qttools to have a `gui` feature.

dg0yt replied:
qt5-tools includes command-line host tools (lupdate, lrelease) as well as GUI target tools (Qt Linguist, Qt Assistant). AFAICT the target tools are barely usable on Android.
(There are also libs (Qt Help) which may include GUI features, even if not needed by the command line tools.)

I had a "quick" feature at an earlier stage of this PR, and I would be glad to to add it again later, to completely opt-out from the qt5-declarative dependency. But this isn't well supported in Qt5, so it is no longer part of this PR. (It is still visible in the `libs` variable in the portfile.)

Neumann-A replied:
I think you want an additional `widgets` feature in `qt5-base` which would automatically control the builds of the tools here.

Neumann-A commented on ports/qt5-tools/portfile.cmake:
```diff
@@ -1,9 +1,33 @@
+        ${OPTIONS}
 )
 
+if(NOT libs)
+    set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
+    file(REMOVE "${CURRENT_PACKAGES_DIR}/debug")
+endif()
```
What tools are remaining?

dg0yt replied:
E.g. lrelease, lupdate. (Typical host tools.)

Neumann-A commented on ports/qt5-script/vcpkg.json:
```diff
@@ -1,13 +1,44 @@
+    "svg"
+  ],
+  "features": {
+    "gui": {
+      "description": "Build GUI features, including QWidgets",
+      "dependencies": [
+        {
+          "name": "qt5-base",
+          "default-features": false,
+          "features": [
+            "gui"
+          ]
+        }
+      ]
     },
```
What exactly is gui used for ?

dg0yt replied:
The Qt GUI module and its dependants.
You don't need it for host tools.

Neumann-A replied:
If I use host tools only, everything should already be gated by `qt5-base` not having the `gui` feature.

JackBoosY replied:
Unresolved item.

Neumann-A commented on ports/qt5-declarative/vcpkg.json:
```diff
@@ -1,14 +1,32 @@
+    "quick"
+  ],
+  "features": {
+    "quick": {
```
`qml` instead of `quick`?

dg0yt replied:
The QML language and infrastructure is the core feature of qt5-declarative (Qt Qml module)
Quick is the UI runtime (Qt Quick module).
So the key questions is if there is use of QML without Quick. E.g. the optional QML plugin provided by kf5i18n doesn't  need Quick (apart from interesting linking errors I studied in the past 24h).

Neumann-A replied:
Hmm `qtdeclarative` does have no feature at all. You probably have the same problem as me with how Qt looks for data stuff it needs for the target. `HostData` seems to need to point to the data of the target and not the host. Try comparing the generated `qt.conf` of `qt6` vs `qt5`

> (apart from interesting linking errors I studied in the past 24h).

check if the generated `*.prl` files are correct. If those don't get corrected nothing works in regards to `qmake`.

Neumann-A replied:
you don't need a feature here. The `quick` build is already gated by the gui feature in `qt5-base`.

dg0yt replied:
> you don't need a feature here. The `quick` build is already gated by the `gui` feature in `qt5-base`.

I need the feature when I want to disable quick but use widgets.

dg0yt replied:
> > (apart from interesting linking errors I studied in the past 24h).
> 
> check if the generated `*.prl` files are correct. If those don't get corrected nothing works in regards to `qmake`.

\<OT> 
The problem is: An incorrect sequence of link libraries when linking the x64-linux [!] `.so` [!] kf5i18n QML plugin, with pthreads not appearing late enough. pthreads is a dependency of Core and should always appear late. I guess this mixed static/shared scenario is neither tested nor desired.
Basically, this problem is independent of my changes.
The prl files are okay. The funny things start in CMake. The `Qt*Config.cmake` file creates library paths with a double `/` between vcpkg `libs` dir and the remaining path. (Upstream behaviour.) This seems to confuse CMake. When I adjust the prefixes in the `Qt*Config.cmake`, the problem goes away.
\</OT>

Neumann-A replied:
```
qtHaveModule(gui):qtConfig(qml-animation) {
    SUBDIRS += quick

    qtConfig(quick-path): \
        SUBDIRS += quickshapes

    qtConfig(testlib): \
        SUBDIRS += qmltest

    qtConfig(quick-particles): \
        SUBDIRS += particles
    qtHaveModule(widgets): SUBDIRS += quickwidgets
}
```

`qtConfig(qml-animation)` should give you a different leverage over the current approach so I won't approve this change here.

JackBoosY replied:
Unresolved item.

Neumann-A commented:
I think you should add a `widgets` feature to `qt5-base` to control the remaining downstream ports. If you want to then control the tools in `qt5-tools` do that in a separate PR only controlling the tools and nothing more.

So what I would like to see is only changes to `qt5-base` here.

Neumann-A commented on ports/qt5-base/portfile.cmake:
```diff
@@ -230,48 +232,68 @@ set(FREETYPE_DEBUG_ALL "${FREETYPE_DEBUG} ${BZ2_DEBUG} ${LIBPNG_DEBUG} ${ZLIB_DE
 x_vcpkg_pkgconfig_get_modules(PREFIX harfbuzz MODULES harfbuzz LIBRARIES)
```
you can probably simplify a lot of the code by using `x_vcpkg_pkgconfig_get_modules`

dg0yt commented:
I could revise the focus of the PR to qt5-base (widgets, icu) and add more of the pkgconfig and maintainability changes.
Then I would follow with a PR fixing mingw builds at least for qt5-base, qt5-imageformats.
And then other PRs could gradually improve the remaining qt5 ecosystem.

dg0yt commented:
> I think you should add a `widgets` feature to `qt5-base` to control the remaining downstream ports.

I just have to avoid an installation order problem. Some port have only transitive dependencies on `qt5-base` now. In the future, they would need a direct dependency on the gate, `qt5-base[core,widgets]`.

Neumann-A commented:
> I just have to avoid an installation order problem. Some port have only transitive dependencies on qt5-base now. In the future, they would need a direct dependency on the gate, qt5-base[core,widgets].

So long you make it the default-feature until downstream qt5 ports do it correctly it doesn't matter. Fixing that still can be done later.

JackBoosY commented:
Please temporary close this PR if you don't have spare time to finish it.

dg0yt commented:
Touching this PR again:
It is not enough to control `gui` or `widgets` only in `qt5-base`: This doesn't cut port dependencies of other qt5 modules.
- To avoid transitively requiring these dependencies for the host, `qt5-tools` really needs the `gui` feature.
- For users who need Qt Widgets also for Qt Tools, but don't need Qt Quick, the `quick` feature of `qt5-declarative` is a necessary and sufficient control. However, this should be a default feature.

dg0yt committed:
Update versions

dg0yt commented:
Requires review, not author-response.

dg0yt commented:
If I gate everything only by the `qt5-base[gui]` feature, it means that for example `vcpkg install qt5-svg` will be empty and *not* install Qt5 SVG if only `qt5-base[core]` is installed. Okay?

dg0yt commented:
> If I gate everything only by the `qt5-base[gui]` feature, it means that for example `vcpkg install qt5-svg` will be empty and _not_ install Qt5 SVG if only `qt5-base[core]` is installed. Okay?

I would like to keep the feature design as proposed in this PR, with `qt5-base[gui]`, `qt5-declarative[quick]` and `qt5-tools[gui]`.
Gating only on `qt5-base[gui]` doesn't allow to cut dependencies from depending ports (e.g. `qt5-imageformats`).
If there are concerns about how the feature selection is passed into each port in absence of direct options, I could still add patches for these options.

JackBoosY commented:
I think you should make agreement with @Neumann-A first.

dg0yt commented:
> I think you should make agreement with @Neumann-A first, sorry for forgot this PR.

Waiting for signal from @Neumann-A.

Neumann-A commented:
> Waiting for signal from @Neumann-A.

I still don't get why you would want to separate qt5-base gui from qt5-tools gui. I mean for qml/quick stuff that makes sense because it drags in the dependency of qt5-declarative but if qt5-base is already build with gui why do you require qt5-tools to be not build wit gui? For me this is just a step into feature hell. 

> If I gate everything only by the qt5-base[gui] feature, it means that for example vcpkg install qt5-svg will be empty and not install Qt5 SVG if only qt5-base[core] is installed. Okay?

This means qt5-svg is dependent on the gui feature and need to add it as a dependency. 

> Gating only on qt5-base[gui] doesn't allow to cut dependencies from depending ports (e.g. qt5-imageformats).

What dependencies are you trying to cut? (i don't think it is qt5-imageformats which probably requires the gui feature?)

Neumann-A commented on ports/qt5-declarative/portfile.cmake:
```diff
@@ -1,5 +1,14 @@
+vcpkg_list(SET OPTIONS)
+if(NOT "quick" IN_LIST FEATURES)
+    vcpkg_list(APPEND OPTIONS "QT.gui.name=" "QT.widgets.name=")
+endif()
```
Influencing `qtConfig(qml-animation)` seems to be the correct way instead of manipulating `qtHaveModule(gui)`

dg0yt replied:
IMO `qml-animation` is to control QML animation features, not QML GUI features in general.

Neumann-A replied:
Yeah that might be your understand of the variable. I just look at 
https://github.com/qt/qtdeclarative/blob/104eae5b17b0ec700391e9539ee3a4f638588194/src/src.pro#L19-L31 and see `qtConfig(qml-animation)` controlling it all (meaning directly influencing `qtHaveModule(quick)` by not adding the quick subdir.)

Neumann-A replied:
That being said that for **that block** controlling `qml-animation` is the same as `"QT.gui.name="` but I haven't checked the rest of the build script to see if `"QT.gui.name="` and `"QT.widgets.name="` have **more side** effects than that.

JackBoosY commented:
Any conclusion here?

dg0yt commented:
> Any conclusion here?

I feel that some requested changes to the current state are not really acceptable or sustainable.
- Controlling features in `qt5-tools` by presence of features in `qt5-base`.
- Controlling features in `qt5-declarative` by properties which correlate but semantically mean a different thing.

I'd appreciate confirmation in one or another direction.

dg0yt commented:
> I'd appreciate confirmation in one or another direction.

Ping.

JackBoosY commented:
I think we might need to confirm these points with @Neumann-A .

dg0yt commented:
> > Any conclusion here?
> 
> I feel that some requested changes to the current state are not really acceptable or sustainable.
> 
>     * Controlling features in `qt5-tools` by presence of features in `qt5-base`.
> 
>     * Controlling features in `qt5-declarative` by properties which correlate but semantically mean a different thing.
> 
> 
> I'd appreciate confirmation in one or another direction.

Ping.

Neumann-A commented:
> Controlling features in qt5-tools by presence of features in qt5-base.

@Adela0814 That is something you guys have to decide if you want ALL the features. I don't see a reason to build `qt5-base` with gui and `qt5-tools` without it.  

```
    vcpkg_list(APPEND OPTIONS
        "QT.quick.name="
        "config.qttools.features.assistant.disable=true"
        "config.qttools.features.designer.disable=true"
        "config.qttools.features.distancefieldgenerator.disable=true"
        "config.qttools.features.pixeltool.disable=true"
        "config.qttools.features.qtdiag.disable=true"
```

Ah maybe I am just annoyed by the name of the feature because it actually doesn't use `"QT.gui.name="` (but that is similar to what i said in https://github.com/microsoft/vcpkg/pull/23668#discussion_r856060692; maybe just go with the qtdeclarative approach and spin out the tools with gui as gui-tools instead and the cmd list tools just as single feature).
Or just follow https://code.qt.io/cgit/qt/qttools.git/tree/configure.json?h=v5.15.7-lts-lgpl and we don't have to discuss at all. (even if it adds 16 features where probably 3-5 are platform dependent and doesn't need to be features explicitly.)

dg0yt commented:
> Controlling features in `qt5-tools` by presence of features in `qt5-base`.

I take the approval of #29153 as an indication that we don't want that.
- There: control qttools dbus by qtbase dbus presence.
- Here: control qt5-tools gui by qt5-base gui presence.

I admit that gui is a cross-cutting concern, and dbusviewer isn't.

Neumann-A commented:
> I take the approval of https://github.com/microsoft/vcpkg/pull/29153 as an indication that we don't want that.

@dg0yt 

> Or just follow https://code.qt.io/cgit/qt/qttools.git/tree/configure.json?h=v5.15.7-lts-lgpl and we don't have to discuss at all. (even if it adds 16 features where probably 3-5 are platform dependent and doesn't need to be features explicitly.)

https://github.com/microsoft/vcpkg/pull/29153 is exactly that. `upstream feature == vcpkg feature`

Adela0814 commented:
Closing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.
